File: Problem1_FirstAndLastPosition.py
# Time Complexity: O(log n)
# Space Complexity: O(1)
# Did this code successfully run on Leetcode: Yes
# Any problem you faced while coding this: No

# Solution:

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if not nums or len(nums) == 0:
            return [-1, -1]

        if target < nums[0] or target > nums[len(nums) - 1]:
            return [-1, -1]

        first = self.binary_search_first(nums, target)
        if first == -1:
            return [-1, -1]

        last = self.binary_search_last(nums, target, first)

        return [first, last]

    # Returns the first index of target
    def binary_search_first(self, nums: List[int], target: int) -> int:
        low, high = 0, len(nums) - 1

        while low <= high:
            mid = low + (high - low) // 2

            if target == nums[mid]:
                if mid == 0 or nums[mid - 1] < nums[mid]:
                    return mid
                elif nums[mid - 1] == nums[mid]:
                    high = mid - 1

            elif target > nums[mid]:
                low = mid + 1

            else:
                high = mid - 1

        return - 1

    # Returns the last index of target
    def binary_search_last(self, nums: List[int], target: int, low: int) -> int:
        high = len(nums) - 1

        while low <= high:
            mid = low + (high - low) // 2

            if target == nums[mid]:
                if mid == len(nums) - 1 or nums[mid + 1] > nums[mid]:
                    return mid
                elif nums[mid + 1] == nums[mid]:
                    low = mid + 1

            elif target > nums[mid]:
                low = mid + 1

            else:
                high = mid - 1

        return - 1

File: Problem1-FirstAndLastPosition.java

// Time Complexity : 2 log n which is equal to  O(logn) where n is number of elements in array
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


class Solution {
    public int[] searchRange(int[] nums, int target) {
        int low = 0 , high = nums.length-1, first=-1, last=-1;
        while(low<=high)   // First binary search to find start point
         {    
            int mid= low + (high-low)/2;
            if(nums[mid]==target)
             { 
               if(mid==0)
                {first = mid;break;}
              else if(nums[mid-1]==nums[mid])
                   high=mid-1;
               else 
                {first=mid; break; }
                
             }
            else if(nums[mid]>target)
                  high=mid-1;
            else 
                low =mid+1;
            
         }
        
        if(first==-1) 
          { 
            return new int[]{first,last};
          } 
        low =first; high=nums.length-1;
        
          while(low<=high)   // Second Binary Search to find end point
          {    
            int mid= low + (high-low)/2;
            if(nums[mid]==target)
              { 
               if(mid==nums.length-1)
                {last = mid;break;}
              else if(nums[mid+1]==nums[mid])
                   low=mid+1;
               else 
                {last=mid; break; }
                
               }
            else if(nums[mid]>target)
                  high=mid-1;
            else 
                low =mid+1;
            
           }
        
        return new int[]{first,last};
        
    }
}